src/VirtoCommerce.CustomerModule.Data/Services/MemberService.cs

```diff
@@ -165,82 +142,150 @@ public virtual async Task SaveChangesAsync(Member[] members)
+        }
 
-                    await repository.RemoveMembersByIdsAsync(members.Select(m => m.Id).ToArray());
-                    await repository.UnitOfWork.CommitAsync();
-                    ClearCache(members);
+        protected virtual void ClearCache(IEnumerable<Member> members)
+        {
+            var models = members as Member[] ?? members.ToArray();
+            ClearSearchCache(models);
 
-                    await _eventPublisher.Publish(new MemberChangedEvent(changedEntries));
-                }
+            foreach (var member in models.Where(x => !x.IsTransient()))
+            {
+                CustomerCacheRegion.ExpireMemberById(member.Id);
             }
         }
 
-        protected virtual void ClearCache(IEnumerable<Member> members)
+        protected virtual void ClearSearchCache(IList<Member> models)
         {
             CustomerSearchCacheRegion.ExpireRegion();
+        }
+
+        protected virtual IList<Member> ProcessModels(IList<MemberEntity> entities, string responseGroup)
+        {
+            return entities?
+                .Select(x =>
+                {
+                    var model = ToModel(x);
+                    return model is null ? null : ProcessModel(responseGroup, x, model);
+                })
+                .Where(x => x is not null)
+                .ToArray();
+        }
 
-            foreach (var member in members.Where(x => !x.IsTransient()))
+        protected virtual Member ToModel(MemberEntity entity)
+        {
+            var model = AbstractTypeFactory<Member>.TryCreateInstance(entity.MemberType);
+            if (model is null)
             {
-                CustomerCacheRegion.ExpireMemberById(member.Id);
+                return null;
             }
+
+            entity.ToModel(model);
+            return model;
+        }
+
+        /// <summary>
+        /// Post-read processing of the model instance.
+        /// A good place to make some additional actions, tune model data.
+        /// Override to add some model data changes, calculations, etc...
+        /// </summary>
+        protected virtual Member ProcessModel(string responseGroup, MemberEntity entity, Member model)
+        {
+            model.ReduceDetails(responseGroup);
+            return model;
+        }
+
+        protected virtual MemberEntity FromModel(Member model, PrimaryKeyResolvingMap keyMap)
+        {
+            var memberEntityType = AbstractTypeFactory<Member>.AllTypeInfos
+                .Where(t => t.MappedType != null && t.IsAssignableTo(model.MemberType))
+                .Select(t => t.MappedType).FirstOrDefault();
+            ArgumentNullException.ThrowIfNull(memberEntityType, nameof(model));
+
+            var dataSourceMember = AbstractTypeFactory<MemberEntity>.TryCreateInstance(memberEntityType.Name);
+            ArgumentNullException.ThrowIfNull(dataSourceMember, nameof(model));
```

Comment:
`ArgumentNullException` is not correct here because `model` is not null.
I think it should be `InvalidOperationException`.